File: solutions/swift/list-ops/1/Sources/ListOps/ListOps.swift
class ListOps {
    class func append<T>(_ list1: [T], _ list2: [T]) -> [T] {
        return list1 + list2
    }

    class func concat<T>(_ lists: [T]...) -> [T] {
        return foldLeft(lists, accumulated: [], combine: append)
    }

    class func filter<T>(_ list: [T], predicate: (T) -> Bool) -> [T] {
        var result: [T] = []
        for item in list {
            if predicate(item) {
                result = append(result, [item])
            }
        }
        return result
    }

    class func map<TIn, TOut>(_ list: [TIn], transform: (TIn) -> TOut) -> [TOut] {
        var result: [TOut] = []
        for item in list {
            result = append(result, [transform(item)])
        }
        return result
    }

    class func length<T>(_ list: [T]) -> Int {
        return list.count
    }

    class func foldLeft<TIn, TOut>(_ list: [TIn], accumulated: TOut, combine: (TOut, TIn) -> TOut) -> TOut {
        var acc = accumulated
        for item in list {
            acc = combine(acc, item)
        }
        return acc
    }
    
    class func foldRight<TIn, TOut>(_ list: [TIn], accumulated: TOut, combine: (TIn, TOut) -> TOut) -> TOut {
        var acc = accumulated
        for item in reverse(list) {
            acc = combine(item, acc)
        }
        return acc
    }

    class func reverse<T>(_ list: [T]) -> [T] {
        return foldLeft(list, accumulated: [], combine: {
            (acc: [T], item: T) -> [T] in
                [item] + acc
        })
    }
}

File: solutions/csharp/list-ops/1/ListOps.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class ListOps
{
    public static int Length<T>(List<T> input)
        => input.Count;

    public static List<T> Reverse<T>(List<T> input)
    {
        input.Reverse();
        return input;
    }

    public static List<TOut> Map<TIn, TOut>(List<TIn> input, Func<TIn, TOut> map)
        => input.Select(map).ToList();

    public static List<T> Filter<T>(List<T> input, Func<T, bool> predicate)
        => input.Where(predicate).ToList();

    public static TOut Foldl<TIn, TOut>(List<TIn> input, TOut start, Func<TOut, TIn, TOut> func)
        => input.Aggregate(start, func);

    public static TOut Foldr<TIn, TOut>(List<TIn> input, TOut start, Func<TIn, TOut, TOut> func)
    {
        var acc = start;

        foreach (var i in Enumerable.Reverse(input))
        {
            acc = func(i, acc);
        }

        return acc;
    }


    public static List<T> Concat<T>(List<List<T>> input)
        => input.SelectMany(m => m).ToList();

    public static List<T> Append<T>(List<T> left, List<T> right)
    {
        left.AddRange(right);
        return left;
    }
}